source/Cosmos.IL2CPU/IL/Conv_I2.cs

```diff
@@ -24,48 +21,45 @@ public override void Execute(_MethodInfo aMethod, ILOpCode aOpCode)
             var xSource = aOpCode.StackPopTypes[0];
             var xSourceSize = SizeOfType(xSource);
             var xSourceIsFloat = TypeIsFloat(xSource);
-            if (xSourceIsFloat)
+
+            if (xSourceSize > 8)
+            {
+                throw new NotImplementedException("Cosmos.IL2CPU.x86->IL->Conv_I2.cs->Error: StackSize > 8 not supported");
+            }
+
+            if (xSourceSize <= 4)
             {
-                if (xSourceSize == 4)
+                if (xSourceIsFloat)
                 {
                     XS.SSE.MoveSS(XMM0, ESP, sourceIsIndirect: true);
                     XS.SSE.ConvertSS2SIAndTruncate(EAX, XMM0);
+                    XS.MoveSignExtend(EAX, AX);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }
-                else if (xSourceSize == 8)
+                else
+                {
+                    XS.Pop(EAX);
+                    XS.MoveSignExtend(EAX, AX);
+                    XS.Push(EAX);
+                }
+            }
+            else if (xSourceSize <= 8)
+            {
+                if (xSourceIsFloat)
                 {
                     XS.SSE2.MoveSD(XMM0, ESP, sourceIsIndirect: true);
+                    XS.Add(ESP, 4);
                     XS.SSE2.ConvertSD2SIAndTruncate(EAX, XMM0);
                     XS.Set(ESP, EAX, destinationIsIndirect: true);
                 }

```

Comment:
A thing that has never been clear to me: when we say xSourceSize we are in some way guaranteed that in reality the value is Int32 / UInt32 / IntPtr / UIntPtr? Or a struct as this will pass giving not the expected results:

```cs
stuct ImNotAnInt {
    byte a;
    byte b;
    byte c;
    byte d;
}
```

I'm pretty sure that the .Net Jitter has an Enum corresponding with the types defined in ECMA (that is not always are the same of the ones exposed by C#).
For example they are talking of TYP_CHAR vs TYP_USHORT here: https://github.com/dotnet/coreclr/issues/12983